core/src/session.rs

```diff
@@ -155,6 +147,48 @@ impl Session {
             .get_or_init(|| MercuryManager::new(self.weak()))
     }
 
+    pub async fn dealer(&self) -> Result<&Dealer, ()> {
+        self.0
+            .dealer
+            .get_or_try_init(|| {
+                let weak = self.weak();
+                let builder = Builder::new();
+                builder.launch(
+                    move || {
+                        let session = weak.try_upgrade();
+
+                        async move {
+                            let session = if let Some(session) = session {
+                                session
+                            } else {
+                                // Just don't finish and hope that someone will come along and clean up the Dealer.
+                                futures_util::future::pending::<Infallible>().await;
+                                unreachable!()
+                            };
+
+                            // TODO: How to handle errors here?
+
+                            let token = keymaster::get_token(
+                                &session,
+                                "65b708073fc0480ea92a077233ca87bd",
+                                "playlist-read",
+                            )
+                            .await
+                            .unwrap();
+
+                            let mut url: Url = "wss://dealer.spotify.com:443/".parse().unwrap();
+                            url.query_pairs_mut()
+                                .append_pair("access_token", &token.access_token);
+                            url
```

Comment:
What should happen if we aren't able to retrieve a token?

It was done on purpose that this closure can't return an error. What should the dealer do other than calling the closure again and again and again?

So it's probably better to periodically try it again inside of the closure. Even if the closure/future does never return: It won't block the executor, and the dealer task can't do anything without a uri.

If possible, we should probably cache tokens and reuse them if they aren't expired.

Reply:
Depends on the reason for not getting a token I guess. When it's a connection error, you'd want to retry periodically and ideally with an increasing back-off timer to not cause a DoS. In case there's a connection, but an invalid response, the same and try some other servers.

For headless machines I'd prefer to just try indefinitely rather than bailing out after some time.